### PR TITLE
nixos/hadoop: module and package improvements

### PR DESCRIPTION
**Pull request by illustris**
###### Motivation for this change
- Disable openfirewallports by default
- Fix typos in mkenableoption
- Replace a lot of boilerplate in HDFS with parametric code
- HADOOP_CONF_DIR to env so that other packages like spark can pick up the config
- Add hadoop nixos tests as passthrough packages
- Fix nixos tests for hadoop 2 and 3.2
- fix yarn nodemanager on hadoop packages without autoPatchelfHook
- fix native libraries support with Java 8
- Add a hadoop gateway role
- add option to pass JVM flags and environment variables
- add definitions for useful role-specific configs
- add support for cgroup based CPU utilization enforcement

TODO in future PRs:
- add support for cgroup based memory utilization enforcement

This is turning out to be tricky. The default memory limit enforcement system will have to do for now.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by illustris** [comment]
I'll rebase after https://github.com/NixOS/nixpkgs/pull/158613 gets merged

**Comment by Artturin** [comment]
Is this change backwards compatible?
At least release notes will be needed for the firewall change

**Comment by illustris** [comment]
The changes should be backwards-compatible, with the exception of openFirewall. There are some minor changes to the defaults too, but nothing breaking. I'll add that to release notes.

File: nixos/modules/services/cluster/hadoop/hdfs.nix [hunk 1]
```diff
@@ -1,191 +1,191 @@
...
+        "-Dcom.sun.management.jmxremote"
+        "-Dcom.sun.management.jmxremote.port=8010"
+      ];
+    };
+    extraEnv = mkOption{
+      type = with types; attrsOf str;
+      default = {};
+      description = "Extra environment variables for ${serviceName}";
+    };
+  } // (if firewallOption then {
```
Comment (Artturin) [inline nixos/modules/services/cluster/hadoop/hdfs.nix:36]: https://nixos.org/manual/nixpkgs/stable/#sec-functions-library-attrset

`optionalAttrs` can be used here and elsewhere where there's a `if X then {X} else {}`

Comment (illustris) [reply nixos/modules/services/cluster/hadoop/hdfs.nix]: Fixed the two occurrences I found. Anything else?

**Pull request merged**